### PR TITLE
delete_in

### PR DESCRIPTION
**Pull request by bcardarella**
Adds `Kernel.delete_in/2` and `Kernel.delete_in/1`

The two functions will delete the last key in either the key list or
path of a deeply nested object.

### CODE REVIEW

**Comment by lexmag** [comment]
There are several calls that behave unexpectedly:
```elixir
iex(1)> delete_in([], [:foo, :bar])
[foo: nil]
iex(2)> delete_in([], [:foo, :bar, :baz])
** (ArgumentError) could not put/update key :bar on a nil value
    (elixir) lib/access.ex:207: Access.get_and_update/3
```

**Comment by bcardarella** [comment]
@lexmag thank you, I figured there would be edge cases I would have to address. I'll get these in

File: lib/elixir/test/elixir/kernel_test.exs [hunk 1]
```diff
@@ -448,6 +448,35 @@ defmodule KernelTest do
...
+    users = %{"john" => %{age: 27}, "meg" => %{age: 23}}
+
+    assert delete_in(users, ["john", :age]) ==
+           %{"john" => %{}, "meg" => %{age: 23}}
+
+    assert_raise FunctionClauseError, fn ->
+      delete_in(users, [])
+    end
+
+    assert_raise ArgumentError, "could not put/update key \"john\" on a nil value", fn ->
```
Comment (bcardarella) [inline lib/elixir/test/elixir/kernel_test.exs]: This error message is not ideal as it appears the function being called is an update one, not a delete. However, because I am repurposing `get_and_update_in` for part of the `delete_in` function we get back its error messages. I'm happy to try and address if anyone has a suggestion.

**Comment by josevalim** [comment]
Thank you! I have two concerns regarding such a function:

1. This function should never be called for structs

2. Many times we want to delete a value and get it at the same time, so something like pop_in would be more general purpose

1 is really a big concern because by simply having this function there, it is more likely someone will call it passing a struct without having a full view of the consequences. I would rather update the docs for update_in and get_and_update_in showing examples of delete and pop respectively with the note you don't want to remove fields from a struct.

**Comment by bcardarella** [comment]
@josevalim 

> 1. This function should never be called for structs

Ok, no worries. I should be able to raise if it acts upon a struct.

> 2. Many times we want to delete a value and get it at the same time, so something like pop_in would be more general purpose

I can implement a `pop_in` function as well. This function is one that I have implemented on several projects so it would definitely be of use in an of itself.

**Comment by josevalim** [comment]
I like the idea of raising for structs. To be clear, this will always raise for structs today:

    delete_in foo[:bar][:baz]

As the syntax altogether is not supported by structs. As well as this:

    delete_in foo, [:bar, :baz]

The only concern is:

    delete_in foo.bar.baz

If the last element is doing a struct access, we should raise.

Regarding pop_in and delete_in, I wouldn't like to add more functions to Kernel so I would rather have one of them. Given delete_in is basically pop_in |> elem(0), that would be my pick.

Finally, it is also worth mentioning that delete_in does not support anonymous functions and it should, similar to to how get_in and get_and_update_in does. Maybe it does support, but we need to document it as well.

File: lib/elixir/lib/kernel.ex [hunk 2]
```diff
@@ -1810,6 +1810,32 @@ defmodule Kernel do
...
+  function. If the key is a function, it will be invoked
+  as specified in `get_and_update_in/3`.
+
+  ## Examples
+
+      iex> users = %{"john" => %{age: 27}, "meg" => %{age: 23}}
+      iex> delete_in(users, ["john", :age])
+      %{"john" => %{}, "meg" => %{age: 23}}
+
+  In case any entries in the middle returns `nil`,
```
Comment (josevalim) [inline lib/elixir/lib/kernel.ex]: One difference between `pop_in` and `delete_in` is that `delete_in` can afford to not fail if any of the entries return nil. This actually makes me more inclined towards `delete_in` because it offers something `update_in` cannot.

**Comment by bcardarella** [comment]
@josevalim because this function relies on `get_and_update_in` to iterate over the keys I suspect it will be OK with anonymous functions, but I will add tests to confirm and document.

> `pop_in |> elem(0)`

Ok, I can get behind that. I'll move things in that direction.

**Comment by bcardarella** [comment]
For my own notes:

@josevalim requested we hold on `pop_in` for now. Also, requested support for anon functions in the `delete` part of the path:

```elixir
[:foo][:bar][delete_fn]
```

**Comment by josevalim** [comment]
We have four options:

1. "hard"  `delete_in` - like today, raises if nil is found along the way
2. "soft" `delete_in` - won't raise if nil is found along the way because the mission will be accomplished
3. "hard" `pop_in` - fetches and deletes the value, raises if nil is found along the way 
4. "soft" `pop_in` - fetches and deletes the value, does not raise if nil is found along the way

1 and 3 can be easily implemented with `update_in` and `get_and_update_in`. So I am more inclined towards 2 or 4. 4 is more general purpose than 2 although I am not sure if getting nil because a value is missing along the way for pop_in is a good idea.

**Comment by bcardarella** [comment]
For 2 & 4 would you be OK with a `try/rescue` to handle this or would you prefer a different implementation of `get_and_update_in`?

**Comment by josevalim** [comment]
@bcardarella I believe we can implement it without try/rescue.

**Comment by josevalim** [comment]
@bcardarella I believe we can implement it without try/rescue and still relying on get_and_update_in. We just need to abort if the current value is nil.

**Comment by bcardarella** [comment]
@lexmag your concern `delete_in([], [:foo, :bar])` should be fixed. It will now raise an `ArgumentError, could not delete key :bar on a nil value`

**Comment by lexmag** [comment]
I like 2 and 4 ("soft"), these options feels natural alongside existing `_in`s:
we use `get_in` in situations when the presence of whole path isn't needed, and same might be expected from `delete_in`.
Also we keep place for future `!` versions if the need arises. :bowtie:

**Comment by josevalim** [comment]
@lexmag there is no need to use get_in  when the presence of whole path isn't needed because it also works without get_in. `params[foo][bar][baz]` won't fail if an intermediate step is missing. :D

**Comment by lexmag** [comment]
@josevalim not if the function as key involved though. :Ь
Anyway just showing connection between functions.

**Comment by josevalim** [comment]
@lexmag this means you are :+1: on the other issue that unifies them given the function as key?

**Comment by lexmag** [comment]
@josevalim more into :+1: side, trying hard to find but don't really know why we shouldn't support it.

**Comment by josevalim** [comment]
Ok, let's go with 4 (soft pop_in)!

**Comment by bcardarella** [comment]
Updated.

`delete_in` will now do soft deletes. Please see the tests to confirm this is the behavior you want.

**Comment by bcardarella** [comment]
Added confirmations that attempting to `delete_in` on a struct will raise.

File: lib/elixir/lib/access.ex [hunk 3]
```diff
@@ -195,8 +195,10 @@ defmodule Access do
       :error -> nil
     end
 
-    {get, update} = fun.(current_value)
-    {get, :maps.put(key, update, map)}
+    case fun.(current_value) do
+      {:skip, _} -> {:ok, map}
```
Comment (josevalim) [inline lib/elixir/lib/access.ex]: This change unfortunately break a code that would call get_and_update_in and return `:skip` as the "gotten" value. Ideally, we should make this work without this change here.

Comment (josevalim) [inline lib/elixir/lib/access.ex]: This may be a hard problem to crack. Let me know if you want me to jump in and give it a try as well.

Comment (bcardarella) [inline lib/elixir/lib/access.ex]: If you wouldn't mind. :)

File: lib/elixir/test/elixir/access_test.exs [hunk 4]
```diff
@@ -17,16 +17,19 @@ defmodule AccessTest do
   [1, 2, 3] = @mod.flatten([1, [2], 3])
 
   @mod -13
-  -13 = @mod
```
Comment (josevalim) [inline lib/elixir/test/elixir/access_test.exs]: Why was this commented out?

Comment (bcardarella) [inline lib/elixir/test/elixir/access_test.exs]: wasn't meant to be, will push the fix

**Comment by josevalim** [comment]
This has been merged but we still have work left to do. :)

**Comment by josevalim** [comment]
Thank you @bcardarella !  :heart: :green_heart: :blue_heart: :yellow_heart: :purple_heart:

**Comment by vrcca** [comment]
@josevalim @bcardarella Sorry to revive this PR after so long, but I was wondering if we decided to not implement `delete_in/2`? I couldn't find it in Kernel. I've kind of implemented this a couple of times in different companies, and I was wondering if this function would make sense nowadays in stdlib? This function is really useful when transforming maps/json. Thanks!

**Comment by josevalim** [comment]
There is `pop_in`. You can get `delete_in` by discarding the first element. No plans to add a `delete_in` itself.
